File: PMImageGalleryController-ObjectiveCExample/PMImageGalleryController-ObjectiveCExample/ViewController.m
//
//  ViewController.m
//  PMImageGalleryView
//
//  Created by Genex on 3/7/17.
//  Copyright © 2017 PM. All rights reserved.
//

#import "ViewController.h"

#import "AppDelegate.h"

#import "PMImage.h"
#import "PMImageViewerController.h"

#import "CHTCollectionViewWaterfallLayout.h"

@interface ViewController () <UICollectionViewDelegate, UICollectionViewDataSource, CHTCollectionViewDelegateWaterfallLayout, PMImageViewerControllerDelegate>
{
    NSMutableArray *imageArray;
    PMImageViewerController *pmImageGalleryObj;
    
    NSInteger hiddenCellIndex;
}

@property (nonatomic) IBOutlet UILabel *headerLabel;
@property (nonatomic) IBOutlet UIView *headerView;
@property (nonatomic) IBOutlet UICollectionView *collectionView;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
    hiddenCellIndex = INT_MAX;
    imageArray = [[NSMutableArray alloc] init];
    
    for (int i = 1; i < 11; i++) {
        NSString *imageName = [NSString stringWithFormat:@"image%d.jpeg",i];
        [imageArray addObject:[[PMImage alloc] initWithImage:[UIImage imageNamed:imageName] thumbNailImage:nil andImageName:imageName]];
    }
    
    CHTCollectionViewWaterfallLayout *layout = (CHTCollectionViewWaterfallLayout *)self.collectionView.collectionViewLayout;
    layout.columnCount = 2;
    
    pmImageGalleryObj = [[PMImageViewerController alloc] initWithViewController:self];
    pmImageGalleryObj.delegate = self;
    pmImageGalleryObj.overlayBackColor = [UIColor clearColor];
    pmImageGalleryObj.fullScreenBackColor = [UIColor clearColor];
    pmImageGalleryObj.isShareEnabled = NO;
    [self.navigationController setNavigationBarHidden:YES animated:NO];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

#pragma mark - UICollectionView Methods
- (CGFloat)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout*)collectionViewLayout minimumInteritemSpacingForSectionAtIndex:(NSInteger)section {
    return 8;
}

- (CGFloat)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout*)collectionViewLayout minimumLineSpacingForSectionAtIndex:(NSInteger)section {
    return 8;
}

- (UIEdgeInsets)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout*)collectionViewLayout insetForSectionAtIndex:(NSInteger)section {
    return UIEdgeInsetsMake(8, 8, 8, 8);
}

- (CGSize)collectionView:(UICollectionView *)collectionView
                  layout:(UICollectionViewLayout*)collectionViewLayout
  sizeForItemAtIndexPath:(NSIndexPath *)indexPath
{
    UIImage *image = ((PMImage *)[imageArray objectAtIndex:indexPath.row % 10]).image;
    
    float cellWidth = ( [[UIScreen mainScreen] bounds].size.width / 2.0 ) - 12; //Replace the divisor with the column count requirement. Make sure to have it in float.
    
    CGRect boundingRect =  CGRectMake(0, 0, cellWidth, MAXFLOAT);
    CGRect rect = AVMakeRectWithAspectRatioInsideRect(image.size, boundingRect);
    
    CGSize size = CGSizeMake(cellWidth, rect.size.height);
    return size;
}

#pragma mark - UICollectionViewDataSource
- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section
{
    return imageArray.count;
}

- (__kindof UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath
{
    static NSString *reuseIdentifier = @"Cell";
    UICollectionViewCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier:reuseIdentifier forIndexPath:indexPath];
    cell.backgroundColor = [UIColor whiteColor];
    cell.layer.borderWidth=1.0f;
    cell.layer.cornerRadius = 5;
    cell.layer.masksToBounds = YES;
    [[cell layer] setBorderColor:[UIColor lightGrayColor].CGColor];
    
    PMImage *pmImageObj = [imageArray objectAtIndex:indexPath.item % 10];
    
    UIImageView *collectionViewImageView = (UIImageView *)[cell viewWithTag:101];
    collectionViewImageView.image = pmImageObj.image;
    collectionViewImageView.contentMode = UIViewContentModeScaleAspectFit;
    
    UILabel *collectionViewLabel1 = (UILabel *)[cell viewWithTag:102];
    collectionViewLabel1.text = pmImageObj.imageName;
    
    if (indexPath.item == hiddenCellIndex)
        cell.hidden = YES;
    else
        cell.hidden = NO;
    
    return cell;
}

- (void)collectionView:(UICollectionView *)collectionView didSelectItemAtIndexPath:(NSIndexPath *)indexPath
{
    [pmImageGalleryObj showImageAtIndex:indexPath.item fromDataArray:imageArray animated:YES];
}

#pragma mark - PMImageGalleryViewControllerDelegate Methods

- (CGRect)getCellFrameForIndex:(NSInteger)index
{
    NSIndexPath *indexPath = [NSIndexPath indexPathForItem:index inSection:0];
    
    UICollectionViewLayoutAttributes *attributes = [self.collectionView layoutAttributesForItemAtIndexPath:indexPath];
    CGRect cellRect = attributes.frame;
    CGRect cellFrameInSuperview = [self.collectionView convertRect:cellRect toView:[self.collectionView superview]];
    
    return cellFrameInSuperview;
}

- (void)visibleImageIndex:(NSInteger)index
{
    hiddenCellIndex = index;
    [_collectionView reloadData];
}

- (void)didDismissImageGalleryViewAtIndex:(NSInteger)index
{
    hiddenCellIndex = INT_MAX;
    [_collectionView reloadData];
}

@end

File: PMImageGalleryController-SwiftExample/PMImageGalleryController-SwiftExample/ViewController.swift
//
//  ViewController.swift
//  PMImageGalleryController-SwiftExample
//
//  Created by Genex on 3/24/17.
//  Copyright © 2017 PM. All rights reserved.
//

import UIKit
import AVFoundation

class ViewController: UIViewController, UICollectionViewDelegate, UICollectionViewDataSource, UICollectionViewDelegateFlowLayout, PMImageViewerControllerDelegate {
    
    var imageArray = [PMImage]()
    var pmImageGalleryObj = PMImageViewerController()
    
    @IBOutlet weak var collectionView: UICollectionView!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        self.navigationController?.setNavigationBarHidden(true, animated: false)
        
        imageArray = [(PMImage(image: UIImage(contentsOfFile: Bundle.main.path(forResource: "image1", ofType: "jpeg")!), thumbNail: nil, andImageName: "image1")),
                      (PMImage(image: UIImage(contentsOfFile: Bundle.main.path(forResource: "image2", ofType: "jpeg")!), thumbNail: nil, andImageName: "image2")),
                      (PMImage(image: UIImage(contentsOfFile: Bundle.main.path(forResource: "image3", ofType: "jpeg")!), thumbNail: nil, andImageName: "image3")),
                      (PMImage(image: UIImage(contentsOfFile: Bundle.main.path(forResource: "image4", ofType: "jpeg")!), thumbNail: nil, andImageName: "image4")),
                      (PMImage(image: UIImage(contentsOfFile: Bundle.main.path(forResource: "image5", ofType: "jpeg")!), thumbNail: nil, andImageName: "image5")),
                      (PMImage(image: UIImage(contentsOfFile: Bundle.main.path(forResource: "image6", ofType: "jpeg")!), thumbNail: nil, andImageName: "image6")),
                      (PMImage(image: UIImage(contentsOfFile: Bundle.main.path(forResource: "image7", ofType: "jpeg")!), thumbNail: nil, andImageName: "image7")),
                      (PMImage(image: UIImage(contentsOfFile: Bundle.main.path(forResource: "image8", ofType: "jpeg")!), thumbNail: nil, andImageName: "image8")),
                      (PMImage(image: UIImage(contentsOfFile: Bundle.main.path(forResource: "image9", ofType: "jpeg")!), thumbNail: nil, andImageName: "image9")),
                      (PMImage(image: UIImage(contentsOfFile: Bundle.main.path(forResource: "image10", ofType: "jpeg")!), thumbNail: nil, andImageName: "image10"))
        ]
        
        pmImageGalleryObj = PMImageViewerController(viewController: self)
        pmImageGalleryObj.delegate = self
        pmImageGalleryObj.overlayBackColor = UIColor.white
        pmImageGalleryObj.fullScreenBackColor = UIColor.black
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    // MARK - UICollectionView Methods
    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return imageArray.count
    }
    
    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "Cell", for: indexPath)
        cell.backgroundColor = UIColor.white
        cell.layer.borderWidth = 1.0
        cell.layer.cornerRadius = 5
        cell.layer.masksToBounds = true
        cell.layer.borderColor = UIColor.lightGray.cgColor
        
        let pmImage_Obj = imageArray[indexPath.item]
        
        let imageView : UIImageView = cell.viewWithTag(101) as! UIImageView
        imageView.image = pmImage_Obj.image
        
        let imageName : UILabel = cell.viewWithTag(102) as! UILabel
        imageName.text = pmImage_Obj.imageName
        
        return cell
    }
    
    func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
        collectionView.deselectItem(at: indexPath, animated: true)
        
        pmImageGalleryObj.showImage(at: indexPath.item, fromDataArray: imageArray, animated: true)
    }
    
    // MARK - UICollectionViewDelegateFlowLayout
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAt indexPath: IndexPath) -> CGSize {
        let pmImage_Obj = imageArray[indexPath.item]
        let cellWidth = UIScreen.main.bounds.size.width/2 - 12
        
        let boundingRect = CGRect(x: CGFloat(integerLiteral: 0), y: CGFloat(integerLiteral: 0), width: cellWidth, height: CGFloat(MAXFLOAT))
        let rect = AVMakeRect(aspectRatio: pmImage_Obj.image.size, insideRect: boundingRect)
        
        return CGSize(width: cellWidth, height: rect.size.height)
    }
    
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, minimumInteritemSpacingForSectionAt section: Int) -> CGFloat {
        return CGFloat(integerLiteral: 8)
    }
    
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, insetForSectionAt section: Int) -> UIEdgeInsets {
        return UIEdgeInsets(top: 8, left: 8, bottom: 8, right: 8)
    }
    
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, minimumLineSpacingForSectionAt section: Int) -> CGFloat {
        return CGFloat(integerLiteral: 8)
    }
    
    // Mark - PMImageViewerControllerDelegate
    func getCellFrame(for index: Int) -> CGRect {
        let indexPath = NSIndexPath(item: index, section: 0)
        
        let attributes = self.collectionView.layoutAttributesForItem(at: indexPath as IndexPath)
        let cellRect = attributes?.frame;
        let cellFrameInSuperview = self.collectionView.convert(cellRect!, to: self.collectionView.superview)
        
        return cellFrameInSuperview;
    }
    
    func visibleImageIndex(_ index: Int) {
        
    }
    
    func  didDismissImageGalleryView(at index: Int) {
        
    }
    
}

